src/config/ConfigBuilder.ts

```diff
@@ -91,15 +95,85 @@ export class ConfigBuilder {
                 this.effectiveConfig.backupAckToClientEnabled = tryGetBoolean(value);
+            } else if (key === 'security') {
+                this.handleSecurity(value);
             } else {
                 throw new RangeError(`Unexpected config key '${key}' is passed to the Hazelcast Client`);
             }
         }
     }
 
+    private static validateSecurityConfiguration(jsonObject: any): void {
+        if ('security' in jsonObject && 'customCredentials' in jsonObject) {
+            throw new RangeError('Ambiguous security configuration is found. ' +
+                'Use one of \'security\' or \'customCredentials\' elements, not both.')
+        }
+    }
+
+    private handleSecurity(jsonObject: any): void {
+        let isCredentialsSet = false;
+        for (const key in jsonObject) {
+            if (isCredentialsSet) {
+                throw new RangeError('Security configuration may only contain one of the supported credential types. ' +
+                    'Multiple credential types are passed to the Hazelcast Client.');
+            }
+
+            const value = jsonObject[key];
+            if (key === 'usernamePassword') {
+                this.handleUsernamePasswordCredentials(value);
+            } else if (key === 'token') {
+                this.handleTokenCredentials(value);
+            } else if (key === 'custom') {
+                this.effectiveConfig.security.custom = value;
+            } else {
+                throw new RangeError(`Unexpected security config ${key} is passed to the Hazelcast Client`);
+            }
+
+            isCredentialsSet = true;
+        }
+    }
+
+    private handleUsernamePasswordCredentials(jsonObject: any): void {
+        let username: string;
+        let password: string;
+        for (const key in jsonObject) {
+            const value = jsonObject[key];
+            if (key === 'username') {
+                username = tryGetStringOrNull(value);
+            } else if (key === 'password') {
+                password = tryGetStringOrNull(value);
+            } else {
+                throw new RangeError(`Unexpected username password credentials option '${key}' is passed to the Hazelcast Client`)
+            }
+        }
+
+        this.effectiveConfig.security.usernamePassword = new UsernamePasswordCredentialsImpl(username, password);
+    }
+
+    private handleTokenCredentials(jsonObject: any): void {
+        let token: string;
+        let encoding: TokenEncoding;
+        for (const key in jsonObject) {
+            const value = jsonObject[key];
+            if (key === 'token') {
+                token = tryGetString(value);
+            } else if (key === 'encoding') {
+                encoding = tryGetEnum(TokenEncoding, value);
+            } else {
+                throw new RangeError(`Unexpected token credentials option '${key}' is passed to the Hazelcast Client`)
+            }
+        }
+
+        if (token == null) {
+            return;
```

Comment:
Should we return here or throw? `token` is a mandatory field in TokenCredentials

Reply:
Yes, I believe it could be better to throw here